core/src/main/java/tech/pegasys/eth2signer/core/Runner.java

```diff
@@ -127,22 +139,74 @@ public void run() {
       metricsEndpoint.start(vertx);
 
       final AzureKeyVaultSignerFactory azureFactory = new AzureKeyVaultSignerFactory();
-
-      final ArtifactSignerProviderFactory factory =
-          new ArtifactSignerProviderFactory(metricsSystem, vertx, azureFactory);
-
-      final ArtifactSignerProvider blsSignerProvider =
-          factory.createBlsSignerProvider(config.getKeyConfigPath());
-
-      final ArtifactSignerProvider ethSecpSignerProvider =
-          factory.createEthSecpSignerProvider(config.getKeyConfigPath());
-
-      final ArtifactSignerProvider fcSecpSignerProvider =
-          factory.createFilecoinSecpSignerProvider(
-              config.getKeyConfigPath(), config.getFilecoinNetwork());
+      final HashicorpConnectionFactory hashicorpConnectionFactory =
+          new HashicorpConnectionFactory(vertx);
+
+      final BlsArtifactSignerFactory blsArtifactSignerFactory =
+          new BlsArtifactSignerFactory(
+              config.getKeyConfigPath(), metricsSystem, hashicorpConnectionFactory);
+
+      final Secp256k1ArtifactSignerFactory ethSecpArtifactSignerFactory =
+          new Secp256k1ArtifactSignerFactory(
+              hashicorpConnectionFactory,
+              config.getKeyConfigPath(),
+              azureFactory,
+              EthSecpArtifactSigner::new);
+
+      final Secp256k1ArtifactSignerFactory fcSecpArtifactSignerFactory =
+          new Secp256k1ArtifactSignerFactory(
+              hashicorpConnectionFactory,
+              config.getKeyConfigPath(),
+              azureFactory,
+              signer -> new FcSecpArtifactSigner(signer, config.getFilecoinNetwork()));
+
+      final Collection<ArtifactSigner> signers =
+          SignerLoader.load(
+              config.getKeyConfigPath(),
+              "yaml",
+              new YamlSignerParser(
+                  blsArtifactSignerFactory,
+                  ethSecpArtifactSignerFactory,
+                  fcSecpArtifactSignerFactory));
+
+      final Map<SignerTypes, List<ArtifactSigner>> signersByType =
+          signers
+              .parallelStream()
+              .collect(
+                  Collectors.groupingBy(
+                      i -> {
+                        if (i instanceof BlsArtifactSigner) {
+                          return SignerTypes.BLS;
+                        } else if (i instanceof EthSecpArtifactSigner) {
+                          return SignerTypes.ETH;
+                        } else if (i instanceof FcSecpArtifactSigner) {
+                          return SignerTypes.FC_SECP;
+                        } else {
+                          return SignerTypes.FC_BLS;
+                        }
+                      }));
+
+      final DefaultArtifactSignerProvider blsSignerProvider =
+          DefaultArtifactSignerProvider.create(
+              signersByType.get(SignerTypes.BLS) != null
```

Comment:
you can use signersByType.getOrDefault(SignerTypes.BLS, emptyList()) to simplify this a bit and avoid that ternary

Reply:
so many API I don't even think about ... but yes, nicer.